mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,126 @@
+    inner: NonNull<aws_imds_client>,
+}
+
+///Configurations for creating a new [ImdsClient].
+#[derive(Debug, Default)]
+pub struct ImdsClientConfig {
+    /// The structure that can pass into the CRT's functions.
+    inner: aws_imds_client_options,
+
+    /// The [ClientBootstrap] to use to create connection to IMDS.
+    client_bootstrap: Option<ClientBootstrap>,
+
+    /// The [RetryStrategy] to use to reschedule failed requests to IMDS. This is reference counted,
+    /// so we only need to hold onto it until this [ImdsClientConfig] is consumed, at which point the
+    /// client will take ownership.
+    retry_strategy: Option<RetryStrategy>,
+}
+
+impl ImdsClient {
+    /// Create a new [ImdsClient]
+    pub fn new(allocator: &Allocator, config: ImdsClientConfig) -> Result<Self, Error> {
+        auth_library_init(allocator);
+
+        // SAFETY: `config` is moved into the [ImdsClient] on success, so `config.inner` (and the values
+        // inside) are guaranteed to live at least as long as this [ImdsClient] does. `allocator` is
+        // guaranteed to be a valid allocator because of the type-safe wrapper.
+        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
+
+        Ok(Self { inner })
+    }
+
+    /// Make a get instance type request to ec2 hosts using [ImdsClient].
+    pub fn get_instance_type<F>(&self, callback: F) -> Result<i32, Error>
+    where
+        F: FnOnce(Result<OsString, Error>) + 'static,
+    {
+        let callback_closure: OnGetResource = Box::new(callback);
+        let callback_wrapper: Box<ImdsClientGetResourceCallback> =
+            Box::new(ImdsClientGetResourceCallback(Box::new(callback_closure)));
+        let callback_wrapper: &mut ImdsClientGetResourceCallback = Box::leak(callback_wrapper);
+        // callback_wrapper is safe so the raw pointer referring to that is safe as well.
+        let callback_raw_ptr: *mut libc::c_void =
+            callback_wrapper as *mut ImdsClientGetResourceCallback as *mut libc::c_void;
+        let get_resource_callback_fn_ptr: Option<unsafe extern "C" fn(*const aws_byte_buf, i32, *mut libc::c_void)> =
+            Some(imds_client_get_resource_callback);
+
+        // Safety: move the reference where the data store in memory out of box potential leaks memory,
+        // however, the binding callback function will drop it once the corresponding crt function finished the execution.
+        // See `extern "C" fn imds_client_get_resource_callback`.
+        unsafe {
+            let result =
+                aws_imds_client_get_instance_type(self.inner.as_ptr(), get_resource_callback_fn_ptr, callback_raw_ptr);
+            Ok(result)
+        }
+    }
+}
+
+impl Drop for ImdsClient {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` points to a valid aws_imds_client, and when this [ImdsClient] is dropped.
+        unsafe { aws_imds_client_release(self.inner.as_ptr()) };
+    }
+}
```

Comment:
Simplify safety comment?

```suggestion
impl Drop for ImdsClient {
    fn drop(&mut self) {
        // SAFETY: `self.inner` is reference counted by CRT, and valid for lifetime of `ImdsClient`.
        unsafe { aws_imds_client_release(self.inner.as_ptr()) };
    }
}
```